src/dist/mod.rs

```diff
@@ -146,12 +145,80 @@ pub struct PartialToolchainDesc {
 /// 1.55-x86_64-pc-windows-msvc
 #[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
 pub struct ToolchainDesc {
-    // Either "nightly", "stable", "beta", or an explicit version number
-    pub channel: String,
+    pub channel: ToolchainChannel,
     pub date: Option<String>,
     pub target: TargetTriple,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub enum ToolchainChannel {
```

Comment:
This looks great! I had started in the same direction but was a bit afraid of the time needed to get it right -- but would be awesome if you figure it out.

I would just name this `Channel` -- I don't think there are other kinds of channels in rustup that we need to disambiguate with?